CLDC-1688 scheme change mailer

CheshireSwift opened the pull request:
Extracted just the E-mail parts of #1029, updated URL to match #1035.

As per #1029, E-mail template exists and is configured to match designs 🙂

CheshireSwift committed:
CLDC-1688 Affected logs mailer

CheshireSwift committed:
CLDC-1688 Correct update_logs path

CheshireSwift commented on app/controllers/locations_controller.rb:
```diff
@@ -48,8 +48,16 @@ def deactivate_confirm
 
   def deactivate
-    if @location.location_deactivation_periods.create!(deactivation_date: params[:deactivation_date]) && reset_location_and_scheme_for_logs!
+    if @location.location_deactivation_periods.create!(deactivation_date: params[:deactivation_date])
```
This technically removes `reset_location_...` from the conditional, but since it's using `update!` it already either returns true or throws, so behaviour should be identical.

CheshireSwift commented on app/controllers/schemes_controller.rb:
```diff
@@ -49,8 +49,15 @@ def deactivate_confirm
 
   def deactivate
-    if @scheme.scheme_deactivation_periods.create!(deactivation_date: params[:deactivation_date]) && reset_location_and_scheme_for_logs!
+    if @scheme.scheme_deactivation_periods.create!(deactivation_date: params[:deactivation_date])
```
(Same comment as `locations_controller` re: removing `reset_location_and_scheme_for_logs!` from condition.)

jamesrose commented on app/mailers/location_or_scheme_deactivation_mailer.rb:
```diff
@@ -0,0 +1,38 @@
+      DEACTIVATION_TEMPLATE_ID,
+      {
+        log_count:,
+        log_or_logs: log_count == 1 ? "log" : "logs",
```
You can use https://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize directly in the template here :)

CheshireSwift replied:
When you say "in the template"; this is going out to Notify rather than ERB internally. Do we have access to it actually in the Notify template (as configured in the web interface), or do you just mean we can put `log_or_logs: pluralize(log_count, "log")` here?

tl;dr do we have access to these sort of helpers in Notify templates?

kosiakkatrina commented on app/mailers/location_or_scheme_deactivation_mailer.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def counts_by_user(logs)
+    logs.each_with_object(Hash.new(0)) do |log, counts|
+      counts[log.created_by] += 1
```
Is this going to create a hash with a key as an entire user or am I missing something?

CheshireSwift replied:
It is, but James has highlighted an alternate approach for this anyway :)

kosiakkatrina reviewed (approved):
Looks good!

jamesrose commented on app/controllers/locations_controller.rb:
```diff
@@ -48,8 +48,16 @@ def deactivate_confirm
+
       flash[:notice] = deactivate_success_notice
+      LocationOrSchemeDeactivationMailer.new.send_deactivation_mails(
+        logs.to_a,
```
```suggestion
        logs,
```

Why `to_a`?

CheshireSwift replied:
I think at one point during development it made the ergonomics of the tests easier, but that no longer appears to be the case. Happy to remove.

jamesrose commented on app/controllers/schemes_controller.rb:
```diff
@@ -49,8 +49,15 @@ def deactivate_confirm
+        logs.to_a,
+        url_for(controller: "lettings_logs", action: "update_logs"),
+        @scheme.service_name,
+      )
```
```suggestion
      LocationOrSchemeDeactivationMailer.new.send_deactivation_mails(
        logs,
        url_for(controller: "lettings_logs", action: "update_logs"),
        @scheme.service_name,
      )
```

jamesrose commented on app/mailers/location_or_scheme_deactivation_mailer.rb:
```diff
@@ -0,0 +1,38 @@
+    logs.each_with_object(Hash.new(0)) do |log, counts|
+      counts[log.created_by] += 1
+    end
+  end
```
```suggestion
  def send_deactivation_mails(logs, update_logs_url, scheme_name, postcode = nil)
        logs.group(:created_by).count.compact.each do |user, count|
      send_deactivation_mail(user, count, update_logs_url, scheme_name, postcode)
    end
  end

private
```

You can do the whole thing in ActiveRecord :)

jamesrose commented on config/routes.rb:
```diff
@@ -111,6 +111,7 @@
       get "csv-download", to: "lettings_logs#download_csv"
       post "email-csv", to: "lettings_logs#email_csv"
       get "csv-confirmation", to: "lettings_logs#csv_confirmation"
+      get "update-logs", to: "lettings_logs#update_logs"
```
Does this route anywhere? I don't see the corresponding action.

CheshireSwift replied:
It doesn't exist yet; I pulled this from Kat's PR to actually add the page. Without at least adding it to the routing, the URL construction for the E-mail fails. This seemed like the tidiest place to draw the line.

Commit:
CLDC-1688 Review markups

Co-authored-by: James Rose <james@jbpr.net>

Update app/controllers/schemes_controller.rb

Co-authored-by: James Rose <james@jbpr.net>

Update app/controllers/locations_controller.rb

Co-authored-by: James Rose <james@jbpr.net>

kosiakkatrina committed:
Refactor to queue messages

kosiakkatrina committed:
Update test

kosiakkatrina committed:
remove irrelevant test
